File: java/src/main/java/leetcode/L0005LongestPalindromicSubstring.java
package leetcode;

public class L0005LongestPalindromicSubstring {

  public String longestPalindrome(String s) {
    String maxSub = "";
    // dp[i][j]: if s.substring(i, j) is palindrome.
    boolean[][] dp = new boolean[s.length()+1][s.length()+1];
    for (int offset = 0; offset < s.length() + 1; offset++) {
      for (int i = 0; i < s.length() + 1 - offset; i++) {
        int j = i + offset;
        if (offset == 0 || offset == 1) {
          dp[i][j] = true;
        } else if (s.charAt(i) == s.charAt(j-1) && dp[i+1][j-1]) {
          dp[i][j] = true;
        }
        if (dp[i][j] && offset > maxSub.length()) {
          maxSub = s.substring(i, i+offset);
        }
      }
    }
    return maxSub;
  }

}

File: python2/l0005_longest_palindromic_substring.py
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        substr = s[0] if s else ''
        
        # If s[i:j] is palindrome.
        dp = [[False] * (len(s) + 1) for i in range(len(s) + 1)]
        # Initialize 0 and 1 length substring.
        for i in range(len(s) + 1):
            dp[i][i] = True
            dp[i][min(i+1, len(s))] = True

        # Traverse diagonally from 2 length substring and above.
        for offset in range(2, len(s) + 1):
            for i in range(len(s) - offset + 1):
                j = i + offset
                if dp[i+1][j-1] and s[i] == s[j-1]:
                    dp[i][j] = True
                    if j - i > len(substr):
                        substr = s[i:j]

        return substr
